langchain/src/tools/discord.ts

```diff
@@ -0,0 +1,340 @@
+import {
+  Client,
+  TextChannel,
+  GatewayIntentBits,
+  Message,
+  ChannelType,
+} from "discord.js";
+import { getEnvironmentVariable } from "../util/env.js";
+import { Tool } from "./base.js";
+
+/**
+ * A tool for retrieving messages from a discord channel using a bot.
+ * It extends the base Tool class and implements the _call method to
+ * perform the retrieve operation. Requires an bot token which can be set
+ * in the environment variables, and a limit on how many messages to retrieve.
+ * The _call method takes the discord channel ID as the input argument.
+ * The bot must have read permissions to the given channel. It returns the
+ * message content, author, and time the message was created for each message.
+ */
+export class DiscordGetMessagesTool extends Tool {
+  static lc_name() {
+    return "DiscordGetMessagesTool";
+  }
+
+  name = "discord-get-messages";
+
+  description = `a discord tool. useful for reading messages from a discord channel. 
+    input should be the discord channel ID. the bot should have read 
+    permissions for the channel`;
+
+  protected botToken: string;
+
+  protected messageLimit: number;
+
+  client = new Client({
+    intents: [GatewayIntentBits.Guilds, GatewayIntentBits.GuildMessages],
+  });
+
+  constructor(
+    botToken: string | undefined = getEnvironmentVariable("DISCORD_BOT_TOKEN"),
```

Comment:
Gave this a closer look - let's actually make all these tools take a single object parameter to prevent awkward signatures like:

`new DiscordSendMessagesTool(undefined, 15)`

The above would be valid if the user sets an env variable rather than passing a token in.